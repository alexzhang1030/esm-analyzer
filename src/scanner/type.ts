import type { ASTNodeLocation } from '@/types'

interface ScanResultBase {
  source: string
  loc: ASTNodeLocation
}

export interface ScanImportResultDefault extends ScanResultBase {
  type: 'default'
  local: string // local name
}

export interface ScanImportResultNamespace extends ScanResultBase {
  type: 'namespace'
  local: string // local name
}

export interface ScanImportResultImport extends ScanResultBase {
  type: 'import'
  subType: 'id' | 'string'
  isType: boolean
  local: string // local name
  imported: string // imported name
}

export type ScanImportResultItem = ScanImportResultDefault | ScanImportResultNamespace | ScanImportResultImport
