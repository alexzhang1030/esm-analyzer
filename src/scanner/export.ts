import type { ResolveVariableDeclaration } from '@/common'
import { getASTNodeLocation, isExportAllDeclaration, isExportDefaultDeclaration, isExportNamedDeclaration, isIdentifier, resolveVariableDeclarationValue } from '@/common'
import type { ASTNode, ASTNodeLocation, t } from '@/types'
import { loop } from '@/utils'

export type ScanExportType =
  | 'ExportNamedDeclaration'
  | 'ExportAllDeclaration'
  | 'ExportDefaultDeclaration'

export interface ScanExportNamedResult {
  type: 'ExportNamedDeclaration'
  kind: t.VariableDeclaration['kind']
  declarations: {
    name: string
    init: ResolveVariableDeclaration
  }[]
}

export interface ScanExportAllResult {
  type: 'ExportAllDeclaration'
  source: string
}

export interface ScanExportDefaultIdentifierResult {
  type: 'ExportDefaultDeclaration'
  subType: 'Identifier'
  id: string
}

export interface ScanExportDefaultObjectResult {
  type: 'ExportDefaultDeclaration'
  subType: 'ObjectExpression'
  properties: {
    key: string
    value: ResolveVariableDeclaration
  }[]
}

export type ScanExportResult = (
  | ScanExportNamedResult
  | ScanExportAllResult
  | ScanExportDefaultIdentifierResult
  | ScanExportDefaultObjectResult
) & {
  loc: ASTNodeLocation
}

export interface ScanExportConfig {
  includeType?: ScanExportType[]
  excludeType?: ScanExportType[]
}

export function scanExport(node: ASTNode, config?: ScanExportConfig): ScanExportResult | null {
  if (
    !isExportNamedDeclaration(node)
    && !isExportDefaultDeclaration(node)
    && !isExportAllDeclaration(node)
  )
    return null
  if (config?.includeType && !config.includeType.includes(node.type))
    return null
  if (config?.excludeType && config.excludeType.includes(node.type))
    return null
  if (node.type === 'ExportNamedDeclaration') {
    if (!node.declaration || node.declaration.type !== 'VariableDeclaration')
      return null
    const vars: ScanExportNamedResult['declarations'] = []
    loop(node.declaration.declarations, (declaration) => {
      if (!isIdentifier(declaration.id))
        return
      const result = {
        name: declaration.id.name,
        init: resolveVariableDeclarationValue(declaration.init),
      }
      if (!result.init)
        return
      vars.push(result)
    })
    return {
      type: 'ExportNamedDeclaration',
      declarations: vars,
      kind: node.declaration.kind,
      loc: getASTNodeLocation(node),
    }
  }
  else if (node.type === 'ExportAllDeclaration') {
    return {
      type: 'ExportAllDeclaration',
      source: node.source.value,
      loc: getASTNodeLocation(node),
    }
  }
  else {
    if (node.declaration.type === 'Identifier') {
      return {
        type: 'ExportDefaultDeclaration',
        subType: 'Identifier',
        id: node.declaration.name,
        loc: getASTNodeLocation(node),
      }
    }
    else if (node.declaration.type === 'ObjectExpression') {
      return {
        type: 'ExportDefaultDeclaration',
        subType: 'ObjectExpression',
        loc: getASTNodeLocation(node),
        properties: node.declaration.properties.map((property) => {
          if (property.type !== 'ObjectProperty' || !isIdentifier(property.key))
            return null
          return {
            key: property.key.name,
            value: resolveVariableDeclarationValue(property.value),
          }
        }).filter(Boolean) as ScanExportDefaultObjectResult['properties'],
      }
    }
  }
  return null
}
