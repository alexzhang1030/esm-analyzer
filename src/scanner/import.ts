import { getASTNodeLocation, isImportDeclaration } from '@/common'
import type { ASTNode, ASTNodeLocation, t } from '@/types'

import { loop } from '@/utils'

interface ScanResultBase {
  source: string
  loc: ASTNodeLocation
}

export interface ScanImportResultDefault extends ScanResultBase {
  type: 'default'
  local: string // local name
}

export interface ScanImportResultNamespace extends ScanResultBase {
  type: 'namespace'
  local: string // local name
}

export interface ScanImportResultImport extends ScanResultBase {
  type: 'import'
  subType: 'id' | 'string'
  isType: boolean
  local: string // local name
  imported: string // imported name
}

export type ScanImportResultItem = ScanImportResultDefault | ScanImportResultNamespace | ScanImportResultImport

// import bar from 'foo'
function resolveDefaultSpecifier(node: t.ImportDefaultSpecifier) {
  return node.local.name
}

// import * as bar from 'foo'
function resolveNamespaceSpecifier(node: t.ImportNamespaceSpecifier) {
  return node.local.name
}

// import { bar } from 'foo'
function resolveImportSpecifier(node: t.ImportSpecifier) {
  let imported: string
  let subType: ScanImportResultImport['subType'] = 'id'
  const isType = node.importKind === 'type'
  if (node.imported.type === 'Identifier') {
    imported = node.imported.name
  }
  else {
    imported = node.imported.value
    subType = 'string'
  }
  return {
    imported,
    local: node.local.name,
    subType,
    isType,
  }
}

export interface ScanImportConfig {
  includeSource?: string[]
  excludeSource?: string[]
  skipType?: boolean
}

export const defaultConfig: Required<ScanImportConfig> = {
  includeSource: [],
  excludeSource: [],
  skipType: false,
}

export function scanImport(node: ASTNode, config: ScanImportConfig = defaultConfig): ScanImportResultItem[] | null {
  if (!isImportDeclaration(node))
    return null
  const items = node.specifiers
  if (!items.length)
    return null
  const result: ScanImportResultItem[] = []
  const source = node.source.value
  const isType = node.importKind === 'type'
  if (config.includeSource?.length && !config.includeSource.includes(source))
    return null
  if (config.excludeSource?.length && config.excludeSource.includes(source))
    return null
  if (config.skipType && isType)
    return null
  loop(items, (item) => {
    if (item.type === 'ImportDefaultSpecifier') {
      result.push({
        type: 'default',
        source,
        local: resolveDefaultSpecifier(item),
        loc: getASTNodeLocation(item),
      })
    }
    else if (item.type === 'ImportNamespaceSpecifier') {
      result.push({
        type: 'namespace',
        source,
        local: resolveNamespaceSpecifier(item),
        loc: getASTNodeLocation(item),
      })
    }
    else if (item.type === 'ImportSpecifier') {
      const { imported, local, subType, isType: subIsType } = resolveImportSpecifier(item)
      if (config.skipType && subIsType)
        return
      result.push({
        type: 'import',
        loc: getASTNodeLocation(item),
        source,
        imported,
        local,
        subType,
        isType: !isType ? subIsType : isType,
      })
    }
  })
  return result
}
