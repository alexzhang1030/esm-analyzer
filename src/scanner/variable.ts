import type { ResolveVariableDeclaration, VariableType } from '@/common'
import {
  getASTNodeLocation, isIdentifier,
  isVariableDeclaration, resolveVariableDeclarationValue,
} from '@/common'
import type { ASTNode, ASTNodeLocation, t } from '@/types'
import { loop } from '@/utils'

export interface ScanVariableDeclarationResult {
  loc: ASTNodeLocation
  kind: t.VariableDeclaration['kind']
  name: string
  init: ResolveVariableDeclaration
}

export interface ScanVariableDeclarationConfig {
  includeType?: VariableType[]
  excludeType?: VariableType[]
}

export function scanVariableDeclaration(node: ASTNode, config?: ScanVariableDeclarationConfig): ScanVariableDeclarationResult[] | null {
  if (!isVariableDeclaration(node))
    return null
  const result: ScanVariableDeclarationResult[] = []
  loop(node.declarations, (declaration) => {
    // TODO: handle more cases, currently only handle `identifier`
    // TODO: handle delay init, e.g. `var` and `let`
    if (!isIdentifier(declaration.id))
      return
    const value = {
      loc: getASTNodeLocation(declaration),
      kind: node.kind,
      name: declaration.id.name,
      init: resolveVariableDeclarationValue(declaration.init, config),
    }
    if (!value.init)
      return
    result.push(value)
  })
  return result
}
