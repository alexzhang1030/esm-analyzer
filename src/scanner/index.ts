import type { ScanImportConfig, ScanImportResult } from './import'
import { scanImport } from './import'
import type { ScanVariableDeclarationConfig, ScanVariableDeclarationResult } from './variable'
import { scanVariableDeclaration } from './variable'
import type { ScanExportConfig } from './export'
import { scanExport } from './export'
import type { AcceptableLang } from '@/types'
import { loadScanners } from '@/common'

interface ScanConfig {
  import?: ScanImportConfig
  variable?: ScanVariableDeclarationConfig
  export?: ScanExportConfig
}

export function scan(code: string, lang: AcceptableLang, config?: ScanConfig) {
  const [imports, variables, exports] = loadScanners(code, lang, [
    node => scanImport(node, config?.import),
    node => scanVariableDeclaration(node, config?.variable),
    node => scanExport(node, config?.export),
  ])
  return {
    imports,
    variables,
    exports,
  } as {
    imports: ScanImportResult[]
    variables: ScanVariableDeclarationResult[]
    exports: ScanVariableDeclarationResult[]
  }
}

export * from './import'
export * from './variable'
