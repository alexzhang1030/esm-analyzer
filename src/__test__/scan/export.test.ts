import { loadScanner } from '@/scanner'
import { scanExport } from '@/scanner/export'

describe('scan export', () => {
  test('export default', () => {
    const code = `
      export default {
        a: 1,
        b: 2
      }
    `
    const result = loadScanner(code, 'js', scanExport)
    expect(result).toMatchSnapshot()
  })
  test('export named declaration', () => {
    const code = `
      export const a = 1
      export const b = 2
      const c = 3
    `
    const result = loadScanner(code, 'js', scanExport)
    expect(result).toMatchSnapshot()
  })
  test('export named specifier', () => {
    const code = `
      export { a, b } from 'foo'
      const c = 3, d = 4
      export { c, d }
    `
    const result = loadScanner(code, 'js', scanExport)
    expect(result).toMatchSnapshot()
  })
  test('export all', () => {
    const code = `
      export * from 'foo'
      export * from './bar'
    `
    const result = loadScanner(code, 'js', scanExport)
    expect(result).toMatchSnapshot()
  })
})

describe('scan export pass config', () => {
  test('includeType', () => {
    const code = `
      export const a = 1
      export const b = 2
      export default {
        a: 1,
        b: 2
      }
    `
    const result = loadScanner(code, 'js', node => scanExport(node, {
      includeType: ['ExportDefaultDeclaration'],
    }))
    expect(result).toMatchSnapshot()
  })
  test('excludeType', () => {
    const code = `
      export const a = 1
      export const b = 2
      export default {
        a: 1,
        b: 2
      }
    `
    const result = loadScanner(code, 'js', node => scanExport(node, {
      excludeType: ['ExportDefaultDeclaration'],
    }))
    expect(result).toMatchSnapshot()
  })
})
