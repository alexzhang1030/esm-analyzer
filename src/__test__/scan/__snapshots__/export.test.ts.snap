// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scan export > export all 1`] = `
[
  {
    "loc": {
      "end": {
        "column": 25,
        "index": 26,
        "line": 2,
      },
      "start": {
        "column": 6,
        "index": 7,
        "line": 2,
      },
    },
    "source": "foo",
    "type": "ExportAllDeclaration",
  },
  {
    "loc": {
      "end": {
        "column": 27,
        "index": 54,
        "line": 3,
      },
      "start": {
        "column": 6,
        "index": 33,
        "line": 3,
      },
    },
    "source": "./bar",
    "type": "ExportAllDeclaration",
  },
]
`;

exports[`scan export > export default 1`] = `
[
  {
    "loc": {
      "end": {
        "column": 7,
        "index": 58,
        "line": 5,
      },
      "start": {
        "column": 6,
        "index": 7,
        "line": 2,
      },
    },
    "properties": [
      {
        "key": "a",
        "value": {
          "type": "NumericLiteral",
          "value": 1,
        },
      },
      {
        "key": "b",
        "value": {
          "type": "NumericLiteral",
          "value": 2,
        },
      },
    ],
    "subType": "ObjectExpression",
    "type": "ExportDefaultDeclaration",
  },
]
`;

exports[`scan export > export named 1`] = `
[
  {
    "declarations": [
      {
        "init": {
          "type": "NumericLiteral",
          "value": 1,
        },
        "name": "a",
      },
    ],
    "kind": "const",
    "loc": {
      "end": {
        "column": 24,
        "index": 25,
        "line": 2,
      },
      "start": {
        "column": 6,
        "index": 7,
        "line": 2,
      },
    },
    "type": "ExportNamedDeclaration",
  },
  {
    "declarations": [
      {
        "init": {
          "type": "NumericLiteral",
          "value": 2,
        },
        "name": "b",
      },
    ],
    "kind": "const",
    "loc": {
      "end": {
        "column": 24,
        "index": 50,
        "line": 3,
      },
      "start": {
        "column": 6,
        "index": 32,
        "line": 3,
      },
    },
    "type": "ExportNamedDeclaration",
  },
]
`;

exports[`scan export pass config > excludeType 1`] = `
[
  {
    "declarations": [
      {
        "init": {
          "type": "NumericLiteral",
          "value": 1,
        },
        "name": "a",
      },
    ],
    "kind": "const",
    "loc": {
      "end": {
        "column": 24,
        "index": 25,
        "line": 2,
      },
      "start": {
        "column": 6,
        "index": 7,
        "line": 2,
      },
    },
    "type": "ExportNamedDeclaration",
  },
  {
    "declarations": [
      {
        "init": {
          "type": "NumericLiteral",
          "value": 2,
        },
        "name": "b",
      },
    ],
    "kind": "const",
    "loc": {
      "end": {
        "column": 24,
        "index": 50,
        "line": 3,
      },
      "start": {
        "column": 6,
        "index": 32,
        "line": 3,
      },
    },
    "type": "ExportNamedDeclaration",
  },
]
`;

exports[`scan export pass config > includeType 1`] = `
[
  {
    "loc": {
      "end": {
        "column": 7,
        "index": 108,
        "line": 7,
      },
      "start": {
        "column": 6,
        "index": 57,
        "line": 4,
      },
    },
    "properties": [
      {
        "key": "a",
        "value": {
          "type": "NumericLiteral",
          "value": 1,
        },
      },
      {
        "key": "b",
        "value": {
          "type": "NumericLiteral",
          "value": 2,
        },
      },
    ],
    "subType": "ObjectExpression",
    "type": "ExportDefaultDeclaration",
  },
]
`;
